### PR TITLE
input cartesian, spherical and depth coordinate systems in function through in prm

### PR DESCRIPTION
**Pull request by mbweller**
Allows user to input boundary temperature, velocity, and traction in preferred coordinate systems.

### CODE REVIEW

**Comment by tjhei** [comment]
Is this developed in coordination with Menno's "canonical coordinate system" work?

File: include/aspect/boundary_traction/function.h [hunk 1]
```diff
@@ -94,6 +95,12 @@ namespace aspect
          * A function object representing the components of the traction.
          */
         Functions::ParsedFunction<dim> boundary_traction_function;
+
+        /**
+             * The coordinate representation to evaluate the function. Possible
+             * choices are depth, cartesian and spherical.
+             */
```
Comment (gassmoeller) [inline include/aspect/boundary_traction/function.h]: Please fix the indentation of the comment.

File: source/boundary_temperature/function.cc [hunk 2]
```diff
@@ -20,8 +20,10 @@
 
 
 #include <aspect/boundary_temperature/function.h>
+#include <aspect/utilities.h>
 #include <aspect/global.h>
 
+
```
Comment (gassmoeller) [inline source/boundary_temperature/function.cc]: please remove this additional empty line

File: source/boundary_temperature/function.cc [hunk 3]
```diff
@@ -85,6 +108,24 @@ namespace aspect
...
+           */
+          prm.declare_entry ("Coordinate system", "cartesian",
+                             Patterns::Selection ("cartesian|spherical|depth"),
+                             "A selection that determines the assumed coordinate "
+                             "system for the function variables. Allowed values "
+                             "are 'cartesian', 'spherical', and 'depth'. 'spherical' coordinates "
+                             "are interpreted as r,phi or r,phi,theta in 2D/3D "
+                             "respectively with theta being the polar angle. 'depth' "
+                             "will create a function, in which only the first "
+                             "variable is non-zero, which is interpreted to "
```
Comment (gassmoeller) [inline source/boundary_temperature/function.cc]: I would change this to: "... in which only the first parameter is non-zero ...". "Variable" suggests the function only *outputs* one component.

File: source/boundary_temperature/function.cc [hunk 4]
```diff
@@ -136,8 +186,7 @@ namespace aspect
 {
   namespace BoundaryTemperature
   {
-    ASPECT_REGISTER_BOUNDARY_TEMPERATURE_MODEL(Function,
-                                               "function",
+    ASPECT_REGISTER_BOUNDARY_TEMPERATURE_MODEL(Function, "function",
```
Comment (gassmoeller) [inline source/boundary_temperature/function.cc]: please put the second argument on a separate line, as it was before.

File: source/boundary_traction/function.cc [hunk 5]
```diff
@@ -42,10 +43,39 @@ namespace aspect
...
+        }
+      else if (coordinate_system == ::aspect::Utilities::Coordinates::CoordinateSystem::spherical)
+        {
+          const std_cxx11::array<double,dim> spherical_coordinates =
+            aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(p);
+          Point<dim> point;
+
+          for (unsigned int d=0; d<dim; ++d)
+            point[d] = spherical_coordinates[d];
+          for (unsigned int d=0; d<dim; ++d)
```
Comment (gassmoeller) [inline source/boundary_traction/function.cc]: please add an empty line between the two loops

File: source/boundary_traction/function.cc [hunk 6]
```diff
@@ -42,10 +43,39 @@ namespace aspect
...
+          point(0) = depth;
+
+          for (unsigned int d=0; d<dim; ++d)
+            traction[d] = boundary_traction_function.value(point,d);
+
+          return traction;
+        }
+
 
-      return traction;
```
Comment (gassmoeller) [inline source/boundary_traction/function.cc]: lets leave the `return traction` here, and remove it from within the if-else clauses. That is not a strict rule, just my preference (there is a debate about which practice is better).

File: source/boundary_traction/function.cc [hunk 7]
```diff
@@ -70,6 +100,25 @@ namespace aspect
...
+           */
+          prm.declare_entry ("Coordinate system", "cartesian",
+                             Patterns::Selection ("cartesian|spherical|depth"),
+                             "A selection that determines the assumed coordinate "
+                             "system for the function variables. Allowed values "
+                             "are 'cartesian', 'spherical', and 'depth'. 'spherical' coordinates "
+                             "are interpreted as r,phi or r,phi,theta in 2D/3D "
+                             "respectively with theta being the polar angle. 'depth' "
+                             "will create a function, in which only the first "
+                             "variable is non-zero, which is interpreted to "
```
Comment (gassmoeller) [inline source/boundary_traction/function.cc]: same comment as above

File: source/boundary_velocity/function.cc [hunk 8]
```diff
@@ -41,8 +42,32 @@ namespace aspect
...
+        }
+      else if (coordinate_system == ::aspect::Utilities::Coordinates::CoordinateSystem::spherical)
+        {
+          const std_cxx11::array<double,dim> spherical_coordinates =
+            aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(position);
+          Point<dim> point;
+
+          for (unsigned int d=0; d<dim; ++d)
+            point[d] = spherical_coordinates[d];
+          for (unsigned int d=0; d<dim; ++d)
```
Comment (gassmoeller) [inline source/boundary_velocity/function.cc]: add empy line

File: source/boundary_velocity/function.cc [hunk 9]
```diff
@@ -80,6 +105,25 @@ namespace aspect
...
+           */
+          prm.declare_entry ("Coordinate system", "cartesian",
+                             Patterns::Selection ("cartesian|spherical|depth"),
+                             "A selection that determines the assumed coordinate "
+                             "system for the function variables. Allowed values "
+                             "are 'cartesian', 'spherical', and 'depth'. 'spherical' coordinates "
+                             "are interpreted as r,phi or r,phi,theta in 2D/3D "
+                             "respectively with theta being the polar angle. 'depth' "
+                             "will create a function, in which only the first "
+                             "variable is non-zero, which is interpreted to "
```
Comment (gassmoeller) [inline source/boundary_velocity/function.cc]: same as above

File: source/boundary_temperature/function.cc [hunk 10]
```diff
@@ -40,7 +42,28 @@ namespace aspect
...
+
+          return boundary_temperature_function.value(point);
+        }
+      else
+        {
+          const double depth = this->get_geometry_model().depth(position);
+          Point<dim> point;
+          point(0) = depth;
+          return boundary_temperature_function.value(point);
+        }
```
Comment (tjhei) [inline source/boundary_temperature/function.cc]: add
```
else
AssertThrow(false, ExcNotImplemented());
```
here and in all other cases. The reasoning is that if the enum changes at any time in the future, it won't silently generate wrong results.

File: source/boundary_temperature/function.cc [hunk 11]
```diff
@@ -40,7 +42,28 @@ namespace aspect
...
+          const std_cxx11::array<double,dim> spherical_coordinates =
+            aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(position);
+          Point<dim> point;
+
+          for (unsigned int i = 0; i<dim; ++i)
+            point[i] = spherical_coordinates[i];
+
+          return boundary_temperature_function.value(point);
+        }
+      else
```
Comment (tjhei) [inline source/boundary_temperature/function.cc]: please do ``else if (...)`` here to spell out the type.

File: source/boundary_temperature/function.cc [hunk 12]
```diff
@@ -85,6 +108,24 @@ namespace aspect
       {
         prm.enter_subsection("Function");
         {
+          /**
+           * Choose the coordinates to evaluate the maximum refinement level
```
Comment (tjhei) [inline source/boundary_temperature/function.cc]: this comment is not needed, I think.

File: source/boundary_temperature/function.cc [hunk 13]
```diff
@@ -107,22 +148,31 @@ namespace aspect
...
-                      << "with expression\n"
-                      << "\t'" << prm.get("Function expression") << "'"
-                      << "More information about the cause of the parse error \n"
-                      << "is shown below.\n";
-            throw;
-          }
-        min_temperature = prm.get_double ("Minimal temperature");
-        max_temperature = prm.get_double ("Maximal temperature");
+        {
+          if (prm.get ("Coordinate system") == "cartesian")
```
Comment (tjhei) [inline source/boundary_temperature/function.cc]: can you move this into a helper function ``Utilities::Coordinates::string_to_coordinate_system(string)``

File: source/boundary_traction/function.cc [hunk 14]
```diff
@@ -42,10 +43,39 @@ namespace aspect
...
+          Point<dim> point;
+
+          for (unsigned int d=0; d<dim; ++d)
+            point[d] = spherical_coordinates[d];
+          for (unsigned int d=0; d<dim; ++d)
+            traction[d] = boundary_traction_function.value(point,d);
+
+          return traction;
+        }
+      else
```
Comment (tjhei) [inline source/boundary_traction/function.cc]: ``else if (...)``

File: source/boundary_traction/function.cc [hunk 15]
```diff
@@ -42,10 +43,39 @@ namespace aspect
...
+        {
+          const double depth = this->get_geometry_model().depth(p);
+          Point<dim> point;
+          point(0) = depth;
+
+          for (unsigned int d=0; d<dim; ++d)
+            traction[d] = boundary_traction_function.value(point,d);
+
+          return traction;
+        }
```
Comment (tjhei) [inline source/boundary_traction/function.cc]: add the AssertThrow here

File: source/boundary_traction/function.cc [hunk 16]
```diff
@@ -70,6 +100,25 @@ namespace aspect
       {
         prm.enter_subsection("Function");
         {
+          /**
+           * Choose the coordinates to evaluate the maximum refinement level
```
Comment (tjhei) [inline source/boundary_traction/function.cc]: same comment as above ;-)

File: source/boundary_traction/function.cc [hunk 17]
```diff
@@ -85,6 +134,15 @@ namespace aspect
       prm.enter_subsection("Boundary traction model");
       {
         prm.enter_subsection("Function");
+        {
+          if (prm.get ("Coordinate system") == "cartesian")
```
Comment (tjhei) [inline source/boundary_traction/function.cc]: replace by function call

**Comment by mbweller** [comment]
Changes should have been implemented now (added test case and an entry into doc/modules/changes/ has been added. Should be good to recheck now

**Comment by gassmoeller** [comment]
Please remove some of the new files in the test folder. I will come over and explain.

**Comment by gassmoeller** [comment]
The one failing test was fixed on master in the meantime. Good to go.

**Pull request merged**